stripeClient throw JsonSyntaxException instead of StripeException

### Describe the bug

Hi
After upgrading the com.stripe:stripe-java dependency from version 26.9.0 to 26.11.0, a test in our project fails:

The test expects stripeClient to throw a StripeException, but instead, a com.google.gson.JsonSyntaxException is thrown.
It looks like the JsonSyntaxException is thrown directly instead of being encapsulated within a StripeException like in the previous version




### To Reproduce

Create a test:
Configure WireMock to return an invalid JSON response when the stripeClient do a call
a JsonSyntaxException should be thrown

`
java.lang.AssertionError: 
Expecting actual throwable to be an instance of:
  com.stripe.exception.StripeException
but was:
  com.google.gson.JsonSyntaxException: Expected a com.google.gson.JsonObject but was com.google.gson.JsonPrimitive; at path $
	at com.google.gson.internal.bind.TypeAdapters$34$1.read(TypeAdapters.java:1152)
	at com.google.gson.Gson.fromJson(Gson.java:1361)
	at com.google.gson.Gson.fromJson(Gson.java:1262)
	...(93 remaining lines not displayed - this can be changed with Assertions.setMaxStackTraceElementsDisplayed)
`

### Expected behavior

stripeClient should throw a StripeException

### Code snippets

_No response_

### OS

linux

### Java version

21

### stripe-java version

v26.11.0

### API version

2024-06-20

### Additional context

_No response_

--- a/src/main/java/com/stripe/net/LiveStripeResponseGetter.java
+++ b/src/main/java/com/stripe/net/LiveStripeResponseGetter.java
@@ -6,7 +6,6 @@ import com.google.gson.JsonPrimitive;
 import com.google.gson.JsonSyntaxException;
 import com.stripe.Stripe;
 import com.stripe.exception.*;
-import com.stripe.exception.ApiKeyMissingException;
 import com.stripe.exception.oauth.InvalidClientException;
 import com.stripe.exception.oauth.InvalidGrantException;
 import com.stripe.exception.oauth.InvalidScopeException;
@@ -289,22 +288,27 @@ public class LiveStripeResponseGetter implements StripeResponseGetter {
   }
 
   private void handleError(StripeResponse response, ApiMode apiMode) throws StripeException {
-    JsonObject responseBody = ApiResource.GSON.fromJson(response.body(), JsonObject.class);
+    try {
+      /*
+      OAuth errors are JSON objects where `error` is a string. In
+      contrast, in API errors, `error` is a hash with sub-keys. We use
+      this property to distinguish between OAuth and API errors.
 
-    /*
-     OAuth errors are JSON objects where `error` is a string. In
-     contrast, in API errors, `error` is a hash with sub-keys. We use
-     this property to distinguish between OAuth and API errors.
-    */
-    if (responseBody.has("error") && responseBody.get("error").isJsonPrimitive()) {
-      JsonPrimitive error = responseBody.getAsJsonPrimitive("error");
-      if (error.isString()) {
-        handleOAuthError(response);
+      Try to read the response body to see if it must be
+      */
+      JsonObject responseBody = ApiResource.GSON.fromJson(response.body(), JsonObject.class);
+      if (responseBody.has("error") && responseBody.get("error").isJsonPrimitive()) {
+        JsonPrimitive error = responseBody.getAsJsonPrimitive("error");
+        if (error.isString()) {
+          handleOAuthError(response);
+        }
+      } else if (apiMode == ApiMode.V2) {
+        handleV2ApiError(response);
+      } else {
+        handleV1ApiError(response);
       }
-    } else if (apiMode == ApiMode.V2) {
-      handleV2ApiError(response);
-    } else {
-      handleV1ApiError(response);
+    } catch (JsonSyntaxException e) {
+      throw makeMalformedJsonError(response.body(), response.code(), response.requestId(), e);
     }
   }
 

--- a/src/test/java/com/stripe/functional/LiveStripeResponseGetterTest.java
+++ b/src/test/java/com/stripe/functional/LiveStripeResponseGetterTest.java
@@ -1,8 +1,6 @@
 package com.stripe.functional;
 
 import static org.hamcrest.MatcherAssert.assertThat;
-import static org.junit.jupiter.api.Assertions.assertEquals;
-import static org.junit.jupiter.api.Assertions.assertInstanceOf;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
 import static org.junit.jupiter.api.Assertions.assertThrows;
 
@@ -12,7 +10,6 @@ import com.stripe.exception.ApiException;
 import com.stripe.exception.IdempotencyException;
 import com.stripe.exception.StripeException;
 import com.stripe.model.Subscription;
-import com.stripe.model.oauth.OAuthError;
 import com.stripe.net.ApiResource;
 import com.stripe.net.HttpClient;
 import com.stripe.net.HttpHeaders;
@@ -73,15 +70,12 @@ public class LiveStripeResponseGetterTest extends BaseStripeTest {
     StripeResponseGetter srg = new LiveStripeResponseGetter(spy);
     ApiResource.setGlobalResponseGetter(srg);
     StripeResponse response =
-        new StripeResponse(
-            400,
-            HttpHeaders.of(Collections.emptyMap()),
-            "I am not JSON :)");
+        new StripeResponse(400, HttpHeaders.of(Collections.emptyMap()), "I am not JSON :)");
     Mockito.doReturn(response).when(spy).requestWithRetries(Mockito.<StripeRequest>any());
     assertThrows(
         StripeException.class,
         () -> {
           Subscription.retrieve("sub_123");
         });
-      }
+  }
 }

--- a/src/test/java/com/stripe/functional/LiveStripeResponseGetterTest.java
+++ b/src/test/java/com/stripe/functional/LiveStripeResponseGetterTest.java
@@ -1,6 +1,8 @@
 package com.stripe.functional;
 
 import static org.hamcrest.MatcherAssert.assertThat;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertInstanceOf;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
 import static org.junit.jupiter.api.Assertions.assertThrows;
 
@@ -10,6 +12,7 @@ import com.stripe.exception.ApiException;
 import com.stripe.exception.IdempotencyException;
 import com.stripe.exception.StripeException;
 import com.stripe.model.Subscription;
+import com.stripe.model.oauth.OAuthError;
 import com.stripe.net.ApiResource;
 import com.stripe.net.HttpClient;
 import com.stripe.net.HttpHeaders;
@@ -63,4 +66,22 @@ public class LiveStripeResponseGetterTest extends BaseStripeTest {
             });
     assertThat(exception.getMessage(), CoreMatchers.containsString("idempotency"));
   }
+
+  @Test
+  public void testErrorWithJsonSyntaxException() throws Exception {
+    HttpClient spy = Mockito.spy(new HttpURLConnectionClient());
+    StripeResponseGetter srg = new LiveStripeResponseGetter(spy);
+    ApiResource.setGlobalResponseGetter(srg);
+    StripeResponse response =
+        new StripeResponse(
+            400,
+            HttpHeaders.of(Collections.emptyMap()),
+            "I am not JSON :)");
+    Mockito.doReturn(response).when(spy).requestWithRetries(Mockito.<StripeRequest>any());
+    assertThrows(
+        StripeException.class,
+        () -> {
+          Subscription.retrieve("sub_123");
+        });
+      }
 }

